scripts/build-platforms-generate.sh

```diff
@@ -0,0 +1,186 @@
+
+stages:
+  - build       # Cross-platform library compilation, unit tests
+
+image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
+EOF
+
+printf "\n"
+
+# Each test directory has its own job
+for test_dir in tests/**/*/; do
+  test_files=("$test_dir"*.test.ts)
+  if [ ${#test_files[@]} -eq 0 ]; then
+    continue
+  fi
+  # Remove trailing slash
+  test_dir="${test_dir%\/}"
+  # Remove `tests/` prefix
+  test_dir="${test_dir#*/}"
+  cat << EOF
+build:linux $test_dir:
+  stage: build
+  needs: []
+  script:
+    - >
+        nix-shell --run '
+        npm test -- --ci --coverage --runInBand ${test_files[@]};
+        '
+  artifacts:
+    when: always
+    reports:
+      junit:
+        - ./tmp/junit/junit.xml
+      coverage_report:
+        coverage_format: cobertura
+        path: ./tmp/coverage/cobertura-coverage.xml
+  coverage: '/All files[^|]*\|[^|]*\s+([\d\.]+)/'
+EOF
+  printf "\n"
+done
+
+# All top-level test files are accumulated into 1 job
+test_files=(tests/*.test.ts)
+cat << EOF
+build:linux index:
+  stage: build
+  needs: []
+  script:
+    - >
+        nix-shell --run '
+        npm test -- --ci --coverage --runInBand ${test_files[@]};
+        '
+  artifacts:
+    when: always
+    reports:
+      junit:
+        - ./tmp/junit/junit.xml
+      coverage_report:
+        coverage_format: cobertura
+        path: ./tmp/coverage/cobertura-coverage.xml
+  coverage: '/All files[^|]*\|[^|]*\s+([\d\.]+)/'
+EOF
+
+printf "\n"
+
+# Using shards to optimise tests
+# In the future we can incorporate test durations rather than using
+# a static value for the parallel keyword
+
+# Number of parallel shards to split the test suite into
+CI_PARALLEL=2
+
+cat << "EOF"
+build:windows:
+  stage: build
+  needs: []
+EOF
+cat << EOF
+  parallel: $CI_PARALLEL
+EOF
+cat << "EOF"
+  tags:
+    - windows
+  before_script:
+    - mkdir -Force "$CI_PROJECT_DIR/tmp"
+    - .\scripts\choco-install.ps1
+    - refreshenv
+  script:
+    - npm config set msvs_version 2019
+    - npm install --ignore-scripts
+    - $env:Path = "$(npm bin);" + $env:Path
+    - npm test -- --ci --coverage --shard=$CI_NODE_INDEX/$CI_NODE_TOTAL --maxWorkers=50%
+  artifacts:
+    when: always
+    reports:
+      junit:
+        - ./tmp/junit/junit.xml
+      coverage_report:
+        coverage_format: cobertura
+        path: ./tmp/coverage/cobertura-coverage.xml
+  coverage: '/All files[^|]*\|[^|]*\s+([\d\.]+)/'
+EOF
+
+printf "\n"
+
+cat << "EOF"
+build:macos:
+  stage: build
+  needs: []
+EOF
+cat << EOF
+  parallel: $CI_PARALLEL
+EOF
+cat << "EOF"
+  tags:
+    - shared-macos-amd64
+  image: macos-11-xcode-12
+  variables:
+    HOMEBREW_NO_INSTALL_UPGRADE: "true"
+    HOMEBREW_NO_INSTALL_CLEANUP: "true"
+    HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK: "true"
+    HOMEBREW_NO_AUTO_UPDATE: "true"
+  before_script:
+    - mkdir -p "$CI_PROJECT_DIR/tmp"
+    - eval "$(brew shellenv)"
+    - brew install node@16
+    - brew link --overwrite node@16
+    - hash -r
```

Comment:
Same with this to go into the `script` section.

Then just remove the `before_script` entirely as the default `before_script` would be used.